Issue 1070 - Docker Overhaul, Configuration Handling Update, Documentation Overhaul (#1072)
PowerDNS-Admin/PowerDNS-Admin: docs/docker.md

# PowerDNS-Admin packaged by [Azorian Solutions](https://azorian.solutions)

The PowerDNS-Admin is a simple web GUI for managing zone configurations of a PowerDNS Authoritative server.

The PowerDNS-Admin app does NOT modify the PowerDNS Authoritative server database directly. Instead, it communicates with the PDNS server via the built-in HTTP API.

The app does have a database for identity management, access control, and caching which can be hosted in either MySQL or SQLite.

- [PowerDNS-Admin GitHub](https://github.com/ngoduykhanh/PowerDNS-Admin)
- [PowerDNS-Admin Settings](https://github.com/ngoduykhanh/PowerDNS-Admin/blob/master/docs/settings.md)

> - [PowerDNS-Admin Wiki](https://github.com/ngoduykhanh/PowerDNS-Admin/wiki)

Author: I have corrected this.

All documentation will be written with the assumption that you are already reasonably familiar with this ecosystem. This includes container concepts, the Docker ecosystem, and more specifically the product or solution that you are deploying. Simply put, I won't be fluffing the docs with content explaining every detail of what is presented.

## Supported tags

\* denotes an image that is planned but has not yet been released.

### Alpine Linux

- main, latest
- 0.2.4-alpine, alpine, 0.2.4

### Debian Linux

- 0.2.4-debian, debian

## Deploying this image

### App configuration

Configuration of the PowerDNS Admin app may be achieved through a mixture of two approaches. With either approach you choose, you will need to be aware of the various settings available for the app.

> [PowerDNS Admin Settings](https://github.com/ngoduykhanh/PowerDNS-Admin/blob/master/docs/settings.md)

Author: Yes, done.

#### Approach #1

You may pass app settings as environment variables to the container. These environment variables will be automatically inserted into the app.config object. Any environment variable that begins with PDA_ or PDAC_ and is entirely capitalized, will be inserted into the app.config object.

> **One key difference** between the two variable prefixes is that variables prefixed with PDAC_ will have the PDAC_ prefix removed before being inserted into the app.config object.

Reviewer: Can you elaborate ? Where are both variables used ? It is still confusing for me, I guess it will be for many users.

Maybe a concrete use case example of both to illustrate can also be welcome.

Author: I have re-written this section to better address this problem and provide a clear path moving forward. It does contain some notes that if agreed upon, should probably be also copied to a higher level location where you wish to enforce standards.

The following example illustrates a scenario where you want to set the PDA_PDNS_API_KEY app setting in a secure fashion. You have the secret API key stored at /run/secrets/PDA_PDNS_API_KEY.

    PDA_PDNS_API_KEY_FILE=/run/secrets/PDA_PDNS_API_KEY

This would be the equivalent of running:

    export PDA_PDNS_API_KEY=CONTENTS_OF_SECRET_FILE

#### Approach #2

With this approach, you may create traditional PowerDNS Admin config files and map them to an accessible location inside the container. Then you can pass an environment variable to indicate where to load the file from.

For example, say you have mapped a python configuration file to /srv/app/configs/my-config.py in the container. You would add the following environment variable to your deployment configuration:

    FLASK_CONF=/srv/app/configs/my-config.py

**!!! NOTICE !!!** - There is a very specific order in which different sources are loaded. Take careful notice to avoid confusion!

> - The config file **/srv/app/configs/default.py** is loaded.

Reviewer: Can you precise it is empty now.

Author: @jbe-dw can you elaborate? I'm thinking a typo may have caused some confusion in reading your comment.

- The **FLASK_CONF** environment variable is loaded if configured.

> - Any instance specific configuration provided via code instantiation is loaded.

Reviewer: In this sentence, do you mean the file set by FLASK_CONF ?

Author: Yes, that is what I am intending to convey by this statement. I have updated the line to be much more clear about this.

Author: @jbe-dw I'm just now realize that some of what I'm covering in this file is actually related directly to configuration handling of the app (irrelevant of Docker). Would you agree that I should relocate the app related documentation to a different file?

- Settings are loaded individually from any environment variables present.
- If all required MySQL settings are present in the app.config object and no SQLALCHEMY_DATABASE_URI app setting has been configured, the app.config setting will be automatically created from the provided MySQL settings.
  - Otherwise, the SQLALCHEMY_DATABASE_URI app setting is set to "sqlite:////srv/app/pdns.db"

### Deploy with Docker Run

To run a simple container on Docker with this image, execute the following Docker command;

    docker run -d -p 8080:80 -e PDAC_SIGNUP_ENABLED=true ngoduykhanh/powerdns-nameserver

### Deploy with Docker Compose

> 

Reviewer: Can you talk about the available compose files provided here ?

Author: Yes, I am updated the document to include an outline of the purpose of these files.

To run this image using Docker Compose, create a YAML file in a place of your choosing and add the following contents to it;

    version: "3.3"
    services:
      powerdns-admin:
        image: ngoduykhanh/powerdns-admin
        restart: unless-stopped
        environment:
          - PDAC_SALT=ntR8BmMQb7MKTJ5YrgQwZr3pxU3uLG
          - PDAC_SECRET_KEY=eJ92QLuennVStHTj5FVwxz7m4r3Ng3
        ports:
          - "8080:80"

Then execute the following Docker Compose command;

    docker-compose -f /path/to/yaml/file.yml up

**!!! NOTICE !!!** - Make sure to set the environment variable `PDAC_SECRET_KEY` to a long random string ([see here for additional information](https://flask.palletsprojects.com/en/1.1.x/config/#SECRET_KEY))

## Building this image

> If you want to build this image yourself, you can easily do so using the **docker/bin/build-image** command I have included.

Reviewer: ```suggestion
If you want to build this image yourself, you can easily do so using the **docker/bin/build-image** script included in the repository.
```

Author: Great catch, I have updated this mistake. As you can tell, it has been quite some time since I have worked in a team environment. =)
